message_listener.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/leekchan/accounting"
+
+	"github.com/bwmarrin/discordgo"
+)
+
+func messageListener(session *discordgo.Session, message *discordgo.MessageCreate) {
+	if message.Author.ID == session.State.User.ID {
+		return
+	}
+
+	response := understandMessage(message.Content)
+
+	if response != "" {
+		session.ChannelMessageSend(message.ChannelID, response)
+	}
+}
+
+func understandMessage(content string) string {
+	if content == "" || content[:1] != "!" {
+		return ""
+	}
+
+	messageParts := strings.SplitN(content[len(content)-(len(content)-1):], " ", 2)
+	command := messageParts[0]
+	var data string
+	if len(messageParts) > 1 {
+		data = messageParts[1]
+	} else {
+		data = ""
+	}
+
+	switch command {
+	case "help":
+		return help(data)
+	case "ping":
+		return "Pong!"
+	case "ding":
+		return "8=====D"
+	case "bet":
+		return parseBet(data)
+	default:
+		return ""
+	}
+}
+
+func help(data string) string {
+	switch data {
+	case "bet":
+		return "Usage: !bet <amount> <decimal odds> <description of bet>\n" +
+			"Amount can either be in units (eg 3.5u) or $/£/€ (eg $3.50)\n" +
+			"Placing a bet will prompt the bot to check in with you 24h later to see if you won"
+	default:
+		return "BetTrackerBot general help:\n" +
+			"`!help`: Shows this message! Use !help <command\\> to get more specific help for anything below\n" +
+			"`!bet`: Commits a bet to the database\n" +
+			"`!ping`: Pong!"
+	}
+}
+
+func parseBet(data string) string {
+	// Expecting 3 data components: bet size, decimal odds and description
+	dataParts := strings.SplitN(data, " ", 3)
+	if (len(dataParts)) != 3 {
+		return "Usage: !bet <amount> <decimal odds> <description of bet>"
+	}
+
+	var inUnits = false
+	if dataParts[0][len(dataParts[0])-1:] == "u" {
+		inUnits = true
+	}
+	var currencySymbol = []rune(dataParts[0])[0]
+
+	// No negative bet sizes
+	if []rune(dataParts[0])[0] == '-' || []rune(dataParts[0])[1] == '-' {
```

Comment:
Could use `strings.HasPrefix`: https://golang.org/pkg/strings/#HasPrefix
Can strip any leading/trailing whitespace with `strings.TrimSpace()`

Shouldn't need to delve into rune arrays as `strings` are rune aware (unless you attempt to index them directly as you probably found returns the bytes :))